Define how trace/span ids are retrieved (#836)
open-telemetry/opentelemetry-specification: specification/trace/api.md

specification](https://www.w3.org/TR/trace-context/#tracestate-header) for
details on this field.

### Retrieving the `TraceId` and `SpanId`

The API must allow retrieving the `TraceId` and `SpanId`, it can be directly on

> the SpanContext or on the objects (in this case SpanContext should have methods

Reviewer: `SpanContext`?

Author: You mean using ` around the name?

to retrieve the `TraceId` and `SpanId` objects):

> * `TraceIdBase16` - returns the base16 encoded `TraceId`

Reviewer 2: which base64? The RFC also defines "base64url"

https://tools.ietf.org/html/rfc4648#section-5

Author: It is base16 (equivalent with hex) https://tools.ietf.org/html/rfc4648#section-8

Reviewer 2: also, is the returned string supposed to be padded or not?

Author: Rename to hex and made sure I added link to the rfc4648 which clarifies that base16==hex :)

Author: I was waiting for that question @yurishkuro. I don't think we can talk about "padding" because the values is 16-byte array and doing a base16([16]byte) will always return a 32 hex-characters string

Author: Clarified that the results are 32/16-hex-characters string.

specification](https://www.w3.org/TR/trace-context/#tracestate-header) for
details on this field.

### Retrieving the `TraceId` and `SpanId`

> The API must allow retrieving the `TraceId` and `SpanId`, it can be directly on

Reviewer 3: ```suggestion
The API MUST allow retrieving the `TraceId` and `SpanId`, it can be directly on
```

> The API must allow retrieving the `TraceId` and `SpanId`, it can be directly on
> the `SpanContext` or on the objects (in this case `SpanContext` should have
> methods to retrieve the `TraceId` and `SpanId` objects):

Reviewer 4: ```suggestion
The API must allow retrieving the `TraceId` and `SpanId` in the following forms.
They can be exposed directly on the `SpanContext` or on objects that represent each (e.g., `SpanContext` has methods to retrieve the `TraceId` and `SpanId` objects which expose these retrieval methods):
```

Reviewer 4: Mostly was confused by what objects "on the objects" is tried to make it clearer

Reviewer 3: This is still confusing to me.
Maybe just state the whole thing a bit less like an API reference documentation, e.g. "The API MUST allow retrieving the `TraceId` and `SpanId` as byte arrays (of length 16 for trace ID and length 8 for span ID) and lowercase, zero-padded hex strings (of length 32 for trace ID and 16 for span ID)."

I would not specify anything more. For example, I think it should also be valid to provide a free function "GetTraceIdAsHex(SpanContext)" or "GetTraceIdAsHex(TraceId)", which is neither "on" the SpanContext nor the TraceId.

Author: What does it mean to zero-pad the hex? TraceId is 16-byte array so I cannot see how you can get a different result than a 32 hex. I am trying to make sure I don't add wording that implies they are numbers.

> * `TraceIdHex` - returns the [hex encoded](https://tools.ietf.org/html/rfc4648#section-8) `TraceId`.

Reviewer 2: RFC 4648 uses upper-case letters. However, https://www.w3.org/TR/trace-context/#trace-id uses lower-case. I would specify lower-case.

Author: Good catch, the RFC says (but examples use upper-case):
> Base 16 encoding is the standard case-insensitive hex encoding and may be referred to as "base16" or "hex".

I tried to dump so "lowercase". If you have a better alternative let me know.

> * `TraceIdHex` - returns the lowercase [hex encoded](https://tools.ietf.org/html/rfc4648#section-8) `TraceId`.

Reviewer 5: Do we need such amount of details in the API spec? I would imagine it is sufficient that the spec says "IDs are 16/8 byte arrays, there MUST be a way to retrieve the ID", but will not specify how exactly the IDs are represented in-memory because different languages may have different best representations.

Author: We need to ensure there are some getters that return the binary (raw) and the lowercase hex version of these arrays

> The result of this API should always be a 32-hex-character lowercase string.

Reviewer 3: ```suggestion
The result of this API should always be a 32-hex-character lowercase string (left-padded with `0` if required).
```

Author: See my previous comment, this cannot happen.

* `TraceIdBinary` - returns the 16-byte array representation of the `TraceId`.
The result of this API should always be a 16-byte array.
* `SpanIdHex` - returns the lowercase [hex encoded](https://tools.ietf.org/html/rfc4648#section-8) `SpanId`.
The result of this API should always be a 16-hex-character lowercase string.

> * `SpanIdBinary` - returns the 16-byte array representation of the `SpanId`.

Reviewer 3: ```suggestion
* `SpanIdBinary` - returns the 8-byte array representation of the `SpanId`.
```

The result of this API should always be a 8-byte array.

> The API should not expose details about how they are internally stored.

Reviewer 4: Does it make sense to add something like "These are the semantics of the APIs but names should be idiomatic to the language and need not match"?

Reviewer 6: If that's not explicitly stated elsewhere to apply globally then we can implicitly assume these are just the conceptual names for the API anyway.
